Refactor to use Elektra

qwepoizt opened the pull request:
This PR refactors Redshift to use Elektra for configuration management.

# Summary of changes
* Removed Redshift's code for parsing configuration file
* Removed Redshift's code for parsing CLI options
* Configured build process to include Elektra
* Added code to load configuration settings from Elektra (from CLI options or configuration file)#837 

# Known limitations
*  This PR has a limitation regarding the `randr` method of adjusting color temperature: It only affects one `crtc`. This could be fixed rather easily by changing the implementation of [src/gamma-randr.c](https://github.com/jonls/redshift/pull/837/files#diff-2b0175f70e9ba148949f9d3d67d99d8441c800473cf244de9aef72d5ce94b30d).

Commit:
Add libelektra specification that corresponds to currently supported config files

Commit:
Remove deprecated configuration value "transition" from specification

Commit:
Revise naming and hierarchy

Commit:
Change to types that better match possible values

Commit:
Add metadata for required plugins

Commit:
Add default values

Commit:
Add "range" metadata

Commit:
Replace regex for time values by check/date

Commit:
Add "require" metadata

Commit:
Remove prefix "redshift/" from key names

Commit:
Revise TODOs

Commit:
Remove infos/plugins (see https://github.com/jonls/redshift/pull/829#discussion_r674678554)

Commit:
Add explanation and links about specification files

Commit:
Improve naming and hierarchy of keys

Commit:
Add missing example, default and require

Commit:
Add link to documentation for adjustment method

Commit:
Improve hierarchy of keys

Commit:
Merge duplicate values randr/crtc resp. drm/crtc and randr/screen resp. vidmode/screen.

Commit:
Improve descriptions of brightness/day and brightness/night

Commit:
Fix wrong type and check of gamma, gamma/night and gamma/day

Commit:
Add note about environment variables

Commit:
Add CLI arguments to spec file

Commit:
Add opt/arg = required

Commit:
Add newline at end of file.

Commit:
Remove obsolete comment, the parts of lcdprocs spec file have been replaced by my own work and feedback from @markus2330.

Commit:
Improve comments and configuration setting descriptions

Commit:
Add support for single value gamma values and improve descriptions and examples.

Commit:
Add default option "auto" for adjustment/method and provider/location which automatically picks a suitable option

Commit:
Fix type of enum keys

Commit:
Add missing check/type for float keys

Commit:
Add files generated by ./bootstrap and ./configure to .gitignore

Commit:
Add folder generated by IntelliJ IDE to .gitinore

Commit:
Add line feed to end of file

Commit:
Merge branch 'extend-gitignore' into improve-config/elektrify

Commit:
Add libelektra as dependency

Commit:
Extend specification file to include more keys

Commit:
Fix wrong type of key provider/location

Commit:
Add range of legal values for keys temp/day and temp/night

Commit:
Remove config keys brightness and gamma in favor of their day/night variants.

Commit:
Change boolean values from "true", "false" to "1" and "0" as required by Elektra

Commit:
Set default value for lat, lon of location provider manual.

Commit:
Split up [provider/time/dawn] and [provider/time/dusk] into time ranges.

Commit:
Add note on renaming enum values

Commit:
Add opt/arg = none for keys added in 770f2e4dab0edb1d68dce71b38d3ffa38ce67710

Commit:
Improve description of key [gamma/preserve]

Commit:
Add keys to spec for oneshotmanual, showing list and help for methods and providers

Commit:
Replace loading of configuration by usage of Elektra

Commit:
Remove now unused files config-ini.h and config-ini.c

Commit:
Fix typo

Commit:
Add files generated by Elektra to .gitignore

Commit:
Add kdb gen and mount to autotools configuration and contributing documentation

qwepoizt commented:
@markus2330: As we just discussed, I have removed the files generated by Elektra from Git and included them in the autotools configuration.

The PR is now ready for review.

I'd especially appreciate a review of the file `src/elektra/redshift.ini` (I propose to use `git diff 034b2cf2..09dc2841 src/elektra/redshift.ini
` because diffing against master on Github is not helpful, as  `improve-config/specification` was not yet merged).  
Since you are an expert on specification, I'd be interested in your opinion on the recent changes there (e.g. name and place in hierarchy of new keys)!

markus2330 commented on src/Makefile.am:
```diff
@@ -32,6 +32,16 @@ EXTRA_redshift_SOURCES = \
+scripts_DATA = elektra/redshift-conf.mount.sh
+
+KDB ?= kdb
+elektra/redshift-conf.c elektra/redshift-conf.h elektra/redshift-conf.mount.sh: elektra/redshift.ini
```
Something is wrong in with this target when compiled in parallel `-j 5`:

```
make  all-recursive
make[1]: Entering directory '/usr/src/redshift/qwepoizt-redshift'
Making all in src
make[2]: Entering directory '/usr/src/redshift/qwepoizt-redshift/src'
Making all in redshift-gtk
make[3]: Entering directory '/usr/src/redshift/qwepoizt-redshift/src/redshift-gtk'
make[3]: Nothing to be done for 'all'.
make[3]: Leaving directory '/usr/src/redshift/qwepoizt-redshift/src/redshift-gtk'
make[3]: Entering directory '/usr/src/redshift/qwepoizt-redshift/src'
  CC       options.o
  CC       redshift.o
  CC       signals.o
  CC       solar.o
  CC       systemtime.o
kdb gen -F ni=elektra/redshift.ini highlevel "/sw/redshift/#0/current" elektra/redshift-conf initFn=loadConfiguration helpFn=printHelpMessage specloadFn=exitForSpecload embeddedSpec=full;
  CC       gamma-drm.o
options.c:40:10: fatal error: elektra/redshift-conf.h: No such file or directory
   40 | #include "elektra/redshift-conf.h"
      |          ^~~~~~~~~~~~~~~~~~~~~~~~~
compilation terminated.
make[3]: *** [Makefile:763: options.o] Error 1
make[3]: *** Waiting for unfinished jobs....
redshift.c:63:10: fatal error: elektra/redshift-conf.h: No such file or directory
   63 | #include "elektra/redshift-conf.h"
      |          ^~~~~~~~~~~~~~~~~~~~~~~~~
compilation terminated.
make[3]: *** [Makefile:763: redshift.o] Error 1
make[3]: Leaving directory '/usr/src/redshift/qwepoizt-redshift/src'
make[2]: *** [Makefile:851: all-recursive] Error 1
make[2]: Leaving directory '/usr/src/redshift/qwepoizt-redshift/src'
make[1]: *** [Makefile:780: all-recursive] Error 1
make[1]: Leaving directory '/usr/src/redshift/qwepoizt-redshift'
make: *** [Makefile:524: all] Error 2
```

After compiling 3 times it works, so some targets miss a dep to elektra/redshift-conf.h

qwepoizt replied:
Thank you for spotting this! Looking into it I discovered another problem: `make -jN` with N>1 executed up to 3 instances of kdb gen in parallel which sometimes lead to corrupted contents in the output file `elektra/redshift-conf.c` and thus a failing build.

After several hours of troubleshooting I have now finally found a solution for these 2 issues that works reliably!

markus2330 commented on CONTRIBUTING.md:
````diff
@@ -25,10 +25,11 @@ $ ./configure --prefix=$HOME/redshift/root \
 ```
 
-Now, build the files:
+Now, build the files and mount the Elektra specification:
````
Elektra in main README not mentioned. Give a short intro there, e.g. how to set location provider and coordinates.

Also write that e.g. `kdb meta-get /sw/redshift/#0/current/mode description` can be used to get the description.

qwepoizt replied:
Fixed!

markus2330 commented on CONTRIBUTING.md:
````diff
@@ -25,10 +25,11 @@ $ ./configure --prefix=$HOME/redshift/root \
 
 ``` shell
 $ make
+$ APP_PATH=`pwd`/src/redshift sh src/elektra/redshift-conf.mount.sh
````
Write that the script will use sudo internally or execute as root.

qwepoizt replied:
Thanks, fixed!

markus2330 commented on CONTRIBUTING.md:
```diff
@@ -50,6 +51,7 @@ Dependencies
 
 * autotools, gettext
 * intltool, libtool
+* libelektra5
```
Write that also the high-level code generation is needed (if it is in a separate package)

qwepoizt replied:
`kdb gen` is part of the main libelektra5 package, no separate packages are required!

markus2330 commented on src/elektra/redshift.ini:
```diff
@@ -0,0 +1,308 @@
+; Elektra specification file for redshift@490ba2aae9cfee097a88b6e2be98aeb1ce990050 (= redshift v1.12 + some small commits)
+;
+; A specification file describes which configuration settings an application supports, their defaults and their associated CLI options. It also "(...) defines how the configuration settings should be validated" (source: https://www.libelektra.org/tutorials/validate-configuration)
```
Line is very long, please reformat.

qwepoizt replied:
Thanks, fixed!

markus2330 commented on src/elektra/redshift.ini:
```diff
@@ -0,0 +1,308 @@
+; Elektra specification file for redshift@490ba2aae9cfee097a88b6e2be98aeb1ce990050 (= redshift v1.12 + some small commits)
```
Rename this file to `redshift.ni` as ini might be confusing.

qwepoizt replied:
Thanks, fixed!

markus2330 commented on src/elektra/redshift.ini:
```diff
@@ -0,0 +1,308 @@
+description = Show program version.
+default = 0
+example = 1
+opt = V
```
Not working for me, it does not print the version:
```
./src/redshift -V
Trying location provider `geoclue2'...
Using provider `geoclue2'.
Waiting for current location to become available...
Location: 47.31 N, 11.50 E
Period: Daytime
Color temperature: 6500K
Brightness: 1.00
```

qwepoizt replied:
Could it be that you did not execute 
``$ APP_PATH=`pwd`/src/redshift sh src/elektra/redshift-conf.mount.sh`` ?

I can reproduce your reported behavior by unmounting `spec:/sw/redshift/...` and `/sw/redshift/...`.

markus2330 replied:
It is mounted:

```
kdb mount | grep redshift
#> redshift.ecf on /sw/redshift/#0/current with name /sw/redshift/#0/current
#> sw_redshift__0_current.overlay.spec.eqd on spec:/sw/redshift/#0/current with name spec:/sw/redshift/#0/current
```

I use 0.9.7-5.5531, I hope https://github.com/ElektraInitiative/libelektra/pull/4015 is included?

qwepoizt replied:
Unfortunately, I'm having trouble reproducing that when specification is properly mounted. Could you please provide more information?

1. Does `src/redshift -v` (lower-case v -> verbose) and other CLI options like `--fade` work?

2. What is the output of ... 

* `src/redshift --elektra-spec` 
* `src/redshift --help` 
* `kdb ls spec:/sw/redshift/#0/current/` 
* `kdb meta-show spec:/sw/redshift/#0/current/version` 
* `kdb meta-show /sw/redshift/#0/current/version` 
* `kdb get /sw/redshift/#0/current/version` 
?

3. Maybe your build contains a broken specification (see https://github.com/jonls/redshift/pull/837#discussion_r706612067).
Could you please pull my latest changes with the fix and then unmount both mointpoints, `make distclean`, bootstrap, configure, make, execute redshift-conf.mount.sh (and optionally make install) and finally `src/redshift -v`.

> I use 0.9.7-5.5531, I hope ElektraInitiative/libelektra#4015 is included?

`version` and `verbose` are not deeply-nested, so https://github.com/ElektraInitiative/libelektra/pull/4015 is unrelated.

qwepoizt replied:
For reference: I used libelektra@e9d0afa7821b82b44940ffe5bfc271f2c69bc663 for testing (current master).

markus2330 replied:
> src/redshift --help

Also simply prints "No more methods to try."

I'll try the other checks later, probably we should open an issue for that.

qwepoizt replied:
Good idea, please see https://github.com/ElektraInitiative/libelektra/issues/4041

qwepoizt replied:
This problem was solved, see https://github.com/ElektraInitiative/libelektra/issues/4041

markus2330 commented on src/elektra/redshift.ini:
```diff
@@ -0,0 +1,308 @@
+; Elektra specification file for redshift@490ba2aae9cfee097a88b6e2be98aeb1ce990050 (= redshift v1.12 + some small commits)
+;
```
A wrong path is used everywhere (`/sw/redshift/#0/current`): it does not contain an org. It should be /sw/jonls/redshift/#0/current` (or whatever the organisation name is), see https://www.libelektra.org/tutorials/integration-of-your-c-application

qwepoizt replied:
Thanks, fixed!

markus2330 commented on src/elektra/redshift.ini:
```diff
@@ -0,0 +1,308 @@
+
+[fade]
+type = boolean
+description = Whether transitions between color temperatures (e.g. from daytime to nighttime) should happen fast or slow.
```
It does not say if true is fast or slow. Rephrase description or use enum.

qwepoizt replied:
Thanks, fixed with rephrasing!

markus2330 commented on src/elektra/redshift.ini:
```diff
@@ -0,0 +1,308 @@
+type = enum
+check/enum = #4
+check/enum/#0 = continual
+check/enum/#1 = print
```
This prints the location in a format that is not suitable to be used in the config:

```
kdb set user:/sw/redshift/#0/current/provider/location/manual/lat "47.31 N"                                                                                             
#> Sorry, module type issued the error C03200:
#> Validation Semantic: The type 'float' failed to match for 'user:/sw/redshift/#0/current/provider/location/manual/lat' with string '47.31 N'
```

qwepoizt replied:
Well spotted, thanks! I changed the output format to be compatible with the config.

markus2330 commented on src/elektra/redshift.ini:
```diff
@@ -0,0 +1,308 @@
+default = 1.0:1.0:1.0
+example = 0.9
+; Regex ensures format "float" or "float:float:float". Adapted from https://www.regular-expressions.info/floatingpoint.html
+check/validation = ^([0-9]*[\.,]?[0-9]+)(:([0-9]*[\.,]?[0-9]+):([0-9]*[\.,]?[0-9]+))?$
```
Here a check/validation/message is appropriate (that "float" or "float:float:float" is expected).

qwepoizt replied:
Thanks, I added a check/validation/message for all regexes!

markus2330 commented on src/elektra/redshift.ini:
```diff
@@ -0,0 +1,308 @@
+description = Prints the help of the adjustment methods.
+default = 0
+example = 1
+opt/long = help-methods
```
Not working? `./src/redshift --help-methods`

qwepoizt replied:
Please see https://github.com/jonls/redshift/pull/837#discussion_r706612433 , it might be the same cause+fix.

markus2330 commented on src/elektra/redshift.ini:
```diff
@@ -0,0 +1,308 @@
+;=======================================================
+; END Settings used by randr and vidmode
+
+; BEGIN Settings used by drm
```
Is this needed?

qwepoizt replied:
No, thanks for the hint. The fact that some adjustment methods share configuration options is actually evident from redshift's source code. Having it stated here could lead to confusion, I've removed it.

markus2330 commented on src/elektra/redshift.ini:
```diff
@@ -0,0 +1,308 @@
+description = The location longitude. Only applies to location provider "manual".
+check/type = float
+check/range = -180.0-180.0
+; Longitude of berlin:
```
```suggestion
; Longitude of Berlin:
```

qwepoizt replied:
Well spotted, thanks, fixed!

markus2330 commented on src/elektra/redshift.ini:
```diff
@@ -0,0 +1,308 @@
+
+[provider/location/manual/lon]
+type = float
+description = The location longitude. Only applies to location provider "manual".
```
Write that negative coordinates are needed for America.

qwepoizt replied:
Thanks, I improved the description for `lat` and `lon`.

markus2330 commented on src/options.c:
```diff
@@ -283,17 +149,201 @@ find_location_provider(
+
+    // Programm mode
+    ElektraEnumMode mode = elektraGetMode(elektra);
+    if (mode == ELEKTRA_ENUM_MODE_CONTINUAL) {
```
switch so that adding modes gives warnings?

qwepoizt replied:
Good idea, I refactored into switch.

markus2330 commented on src/options.c:
```diff
@@ -283,17 +149,201 @@ find_location_provider(
+        const location_provider_t *provider = find_location_provider(location_providers, locationProviderName);
+        if(provider == NULL) {
+            // User picked a locationProvider provider which is not supported in this build.
+            fprintf(stderr, _("The chosen locationProvider provider is not supported in this build of redshift.\n"));
```
```suggestion
            fprintf(stderr, _("The chosen locationProvider %s is not supported in this build of redshift.\n", provider));
```

qwepoizt replied:
Well spotted, thanks, I fixed that!

markus2330 reviewed (changes requested):
Great work for a first draft: I was able to run and configure redshift!

qwepoizt commented:
@markus2330 Thanks for your elaborate review, I'll work through your remarks!

Commit:
Fix parallel execution of make

Commit:
Update README.md with infos about changing configuration and remove obsolete sample configuration file.

Solves https://github.com/jonls/redshift/pull/837#discussion_r706225769.

Commit:
Add note that mounting specification requires root privileges.

Solves https://github.com/jonls/redshift/pull/837#discussion_r706227664

Commit:
Reformat comment about specification files

Commit:
Change file suffix of specification file to .ni.

Solves https://github.com/jonls/redshift/pull/837#discussion_r706233696

Commit:
Add organisation part "jonls" to redshift's application key name.

Solves https://github.com/jonls/redshift/pull/837#discussion_r706248711

Commit:
Improve description and fix wrong mapping of "fastfade" option.

Solves https://github.com/jonls/redshift/pull/837#discussion_r706249256

Commit:
Ensure string format of configured location is compatible with specification.

Solves https://github.com/jonls/redshift/pull/837#discussion_r706253739

Commit:
Remove obsolete compass direction abbreviations.

Addendum to 5c1b3c23160b8ea8d3c70284170197888f897fa3.

Commit:
Add human-readable error message for options gamma/day and gamma/night.

Solves https://github.com/jonls/redshift/pull/837#discussion_r706268909

Commit:
Remove obsolete comments from specification file.

https://github.com/jonls/redshift/pull/837#discussion_r706272570

Commit:
Fix typo.

Solves https://github.com/jonls/redshift/pull/837#discussion_r706272685

Commit:
Add info about negative values for "lat" and "lon" options.

Solves https://github.com/jonls/redshift/pull/837#discussion_r706273118

Commit:
Refactor mapping of programm mode from if-else to switch.

This triggers compiler warnings when a new mode is added to the specification but not handled in code.

Solves https://github.com/jonls/redshift/pull/837#discussion_r706274940.

Commit:
Improve error messages for unsupported location provider/adjustment method.

Solves https://github.com/jonls/redshift/pull/837#discussion_r706275884

qwepoizt commented:
@markus2330 I've incorporated all your remarks! IMO, after fixing https://github.com/ElektraInitiative/libelektra/issues/4041 this is ready to be merged into jonls/814-improve-config as a first version of elektrified redshift with the stated limitations.

Commit:
Make error messages more user-friendly

Commit:
Improve error messages

Commit:
Improve error message about no working location provider

Commit:
Add line breaks to error messages

Commit:
Improve usability of automatic adjustment method selection:

1. Make all methods eligible for autostart.
2. Move DRM method down in priority list -> it will be tried last before the dummy method.

This improves usability: Users who start redshift in a build that only has "dummy" or "drm" will be informed about the situation. Before this commit, the startup failed when only "dummy" was available, which can lead to confusion.

Commit:
Improve usability of method and location provider listing.

Users are now informed, if their build of redshift contains no method/provider that would work on their system.

Commit:
Improve usability of auto mode for method and location provider.

Commit:
Improve usability of auto mode for method and location provider.

Commit:
Use stderr only for errors, not for infos.

Commit:
Improve info messages for auto mode for method and location provider.

Commit:
Improve info message for location available.

Commit:
Revert CONTRIBUTING.md which was commited too early in 4c7e6636b8ac6ec306ac43f2ce5b1ca63213f1cb.

Commit:
Improve CONTRIBUTING.md:

1. Emphasize that redshift's functionality will be very limited without installed dependencies.
2. Improve structure by adding more headers and enumerations.
3. Add details about purpose and options of ./configure script
4. Emphasize that dependencies are optional but suggested
5. Emhpasize that "-dev" packages are required for dependencies.

Commit:
Move check for version flag to the top of option parsing so redshift can exit early.

Commit:
Implement printing help messagewhen short CLI option "-h" is used.

Previously, only "--help" worked.

Commit:
Change key hierarchy and CLI option name of "fastfade"

markus2330 commented:
@qwepoizt It would be important that the build server passes. Before that we cannot really ask people to have a closer look.

Commit:
Remove dependency to file redshift.conf.sample from Makefile.am and appveyor.yml.

This is an addendum to 20f98909e7ca350429a5103ef334dfa48e120e4e

Commit:
Add copy of redshift.ni in dump format for windows builds

Commit:
Download and install Elektra during appveyor build

Commit:
Replace wget by curl in appveyor.yml script

Commit:
Replace cur by Start-FileDownload in appveyor.yml script

Commit:
Fix typo in appveyor.yml

Commit:
Replace unzip by 7z in appveyor.yml script

Commit:
Fix wrong path in appveyor.yml

Commit:
Fix wrong path in appveyor.yml

Commit:
Rewrite appveyor.yml to support Redshift with Elektra

Commit:
Make files generated by elektra compatible with windows builds

Commit:
Fix AC_DEFINE

Commit:
Fix makefile for windows builds, add elektra resources to zip artefact

Commit:
Remove reference to already deleted file src/config.ini.c from POTFILES.in

Commit:
Replace Redshift's program_mode_t enum by enum generated by Elektra

Commit:
Remove Redshift's validation of temperature and brightness values because they are handled by Elektra

Commit:
Replace ELEKTRA_DOWNLOAD_URL with "master"

Commit:
Merge remote-tracking branch 'qwepoizt/adapt-for-windows' into improve-config/elektrify

Commit:
Add macro branching in options.h

markus2330 commented on configure.ac:
```diff
@@ -32,6 +32,7 @@ esac
 AC_CHECK_TOOL([WINDRES], [windres], [])
 AS_IF([test "x$build_windows" = "xyes" -a -n "x$WINDRES"], [
 	enable_windows_resource=yes
+	AC_DEFINE([WINDOWS_BUILD], [1], [Add a DEFINE that this is a Windows build])
```
It would surprise me very much if autotools do not already have a solution for that.

qwepoizt replied:
Good point, thank you!

I haven't found a more elegant way to accomplish this and I found this solution to have a big benefit:  `WINDOWS_BUILD` will be defined if and only if the rest of the build will also be made Windows-compatible.

E.g. a solution like https://stackoverflow.com/a/38899152 would have most of the build depend on `AC_CHECK_TOOL([WINDRES], [windres], [])` while  `WINDOWS_BUILD` would only be defined if `${host_os}` matches  `mingw*`.

markus2330 commented on src/elektra/windows/redshift-conf.c:
```diff
@@ -0,0 +1,754 @@
+// clang-format off
```
Is there a problem with the code generator under windows? Then it is probably better to either:

1. fix the problem with the code generator
2. always add the generated files (for any platform)

qwepoizt replied:
Thanks for the hint! I have implemented 2. and modified the Makefile and CONTRIBUTING.md so that developers are aware they need to manually trigger the generation on changes of the specification files. I will open an issue in Elektra's repo for compatiblity of the code generator with Windows.

markus2330 reviewed (commented):
Some questions about the windows build.

markus2330 reviewed (approved):
Looking forward to see what others are saying.

Commit:
Add files generated by Elektra to git, document how to regenerate them.

qwepoizt commented:
@jonls I have finished refactoring Redshift to use Elektra.

To test, you can check out the source branch of this PR, [install the libelektra dependency](https://www.libelektra.org/installation/installation) and follow the updated instructions in [CONTRIBUTING.md](https://github.com/jonls/redshift/blob/dcd3ab81632158ed92c3c4cb0a600f72ed33f61b/CONTRIBUTING.md) .

I am looking forwarding to your thoughts and feedback!

zsugabubus commented:
I'm just a nobody from the Internet but does this added 4000 LOC + (random?)
dependency really worth? Currently the whole C source code is about 8000 LOC.
Eh? That is 50% increase for what? Not speaking about the headache for thousands of users caused by throwing out simple, plain text config files and CLI. Project had no commits for more than a year. Return from the ashes but with style. Yeah.

Elektra has more unresolved issues that stars. Not sure how good metric it is but is this really better than editing plain text config files once in a lifetime?

Otherwise I am not clear what is the issue around GTK interface, if any, but it cannot be that bad...

How does this handle multiple $DISPLAYs?

Am I insane?

markus2330 commented:
Thank you for bringing some life in this PR!

> I'm just a nobody from the Internet but does this added 4000 LOC + (random?) dependency really worth?
> Currently the whole C source code is about 8000 LOC.
> Eh? That is 50% increase

Which 4000 LOC are you talking about? This PR **decreases** the LOC of redshift. It adds some documentation and two generated files, maybe this confused you? (Then people do not need to have the code generator installed to compile Redshift.)

> for what?

https://www.libelektra.org/ explains this :wink: 

> Not speaking about the headache for thousands of users caused by throwing out simple, plain text config files and CLI.

Please do not report misinformation. Elektra also uses plain text file configs and command-line arguments.

Where it differs: It adds additional ways to change and specify configuration settings, that were not present in redshift before, e.g., `kdb set user:/sw/redshift/#0/current/mode continual ` allows you to change the mode in the config file.

> Project had no commits for more than a year. Return from the ashes but with style. Yeah.
> Elektra has more unresolved issues that stars. Not sure how good metric it is

:laughing: Yes, very funny metric. Sounds like interpretation of sign of the zodiac based on the first commit.

> but is this really better than editing plain text config files once in a lifetime?

Again: please do not report misinformation. Elektra does not interfere with this habbit, **you** can continue editing plain text config files. For scripts or configuration management tools, however, editing plain text config files is a major pain, that is why Elektra exists.

> Otherwise I am not clear what is the issue around GTK interface, if any, but it cannot be that bad...
> How does this handle multiple $DISPLAYs?

@qwepoizt can you maybe answer these questions?

schuberttobias commented:
@markus2330 Yes, I'll take a look!

qwepoizt commented:
@markus2330 Yes, I'll take a look!

zsugabubus commented:
> It adds some documentation and two generated files, maybe this confused you? (Then people do not need to have the code generator installed to compile Redshift.)

Yes, that was it. Sorry. But installing code generator for developers is that painful? (Will it be easier for users to install their part?)

> Please do not report misinformation. Elektra also uses plain text file configs and command-line arguments.

Sorry again, I had misunderstanding here, as well. So it modifies configuration files directly (if I understand it correctly right now). Then it means I can forget about comments and such?

Anyway, maybe this comment from #814 that made me think it will cause headache for some people:

> backwards compatiblity: I have decided to break backwards compatibility in favor of a revised hierarchy, naming and typing of the values. That makes the file easier to read, use and less error-prone for human editors. As redshift's config file is quite small it should be easy for users to recreate theirs (based on an example) during an upgrade.

Though I still fail to see why in case of Redshift, where I have a <15 line simple, static, plain-text configuration requires an external program to manage it.

Reloading? SIGHUP, if really needed.

Please absolutely do not take it as an offend but Electra currently has 21 issue identified as bug. Some of it is from years ago. When they will be cleaned up? Probably not all of them are critical, okay, but for example "silently deletes child keys" sounds a bit terrifying. I imagine a configuration handling system should be robust. After it, it will not be enough that I have an issue with a program, now I _may_ have to fight with the program I use to configure it.

A new moving part. So many things can go wrong.

> For scripts or configuration management tools, however, editing plain text config files is a major pain, that is why Elektra exists.

Some random thoughts:

* You struggled editing a configuration file but now this new layer of abstraction makes everything magically crystal clear.
* Instead of file paths you get virtual paths to remember and custom commands to operate on them.
* Instead of text editor you get a CLI with lot's of typing boilerplate with shells**t (sorry, escaping) OR a visual interface, where I guess you get a textbox aka. a single-line stupid "text-editor".
* The unformatted description barely satisfying for the more complex cases.
* Is it expected that a user who is unable to open a text file, will "hack" inside a terminal? Providing a convenient interface for non-programmers probably requires more complex GUI elements and layouts. Otherwise a long text file with lot's of comments with one-one config line in between could serve just as well.
* Me, who does a little programming, I prefer to see the complete config file in whole, opened next to the docs and reload the program only after I have finished editing. But it's just me.

**Of course, I may be totally wrong.**

After all, advantages really outweigh disadvantages in case of Redshift?

markus2330 commented:
>  Yes, that was it. Sorry. But installing code generator for developers is that painful? (Will it be easier for users to install their part?)

It normally shouldn't be a problem because by default the code generator gets installed with Elektra. Iirc the reason the generated files are checked in is only because of a problem in building Redshift in Windows.

> Then it means I can forget about comments and such?

No, most plugins, e.g. TOML and hosts, do great effort to keep comments. There are some plugins that discard comments, e.g. JSON.

> Though I still fail to see why in case of Redshift, where I have a <15 line simple, static, plain-text configuration requires an external program to manage it.

The need has little to do with the size of the configuration settings but if there is a reason that external applications might want to change the configuration. I think in Redshift there is a need to change configuration programmatically, e.g., when you change the location.

> Please absolutely do not take it as an offend but Electra currently has 21 issue identified as bug. Some of it is from years ago. When they will be cleaned up? Probably not all of them are critical, okay, but for example "silently deletes child keys" sounds a bit terrifying. I imagine a configuration handling system should be robust. After it, it will not be enough that I have an issue with a program, now I may have to fight with the program I use to configure it.

I cannot argue with that: Elektra is not yet 1.0. We currently focus on getting the API perfect, bugs in tooling unfortunately cannot get too much of our attention. So yes of course early adapters need to struggle with a few annoyances. If you actually hit the `kdb mv` bug, please ping the issue and it will get our attention again. I never hit it, as doing `kdb mv` is something rather rare.

>  Instead of file paths you get virtual paths to remember and custom commands to operate on them.

As said, in Elektra you can still edit config files. `kdb file` gives you the path where the config file is, completely bypassing Elektra. As this is dangerous (you might destroy the configuration and Redshift might not be able to start), Elektra also supports `kdb editor` which starts up your favorite editor with your favorite configuration file format and will only take over configuration iff the config is syntactically and semantically okay (semantics as far as specified in the specification, see src/elektra/redshift.ni in this PR), like with visudo.

> Instead of text editor you get a CLI with lot's of typing boilerplate with shells**t (sorry, escaping) OR a visual interface, where I guess you get a textbox aka. a single-line stupid "text-editor".

The Web-UI gives you buttons according to the type in the specification, e.g., a boolean will give you a checkbox.

> The unformatted description barely satisfying for the more complex cases.

What do you mean?

> Is it expected that a user who is unable to open a text file, will "hack" inside a terminal?

No. Elektra gives you many different UIs, users can choose from them in the same way as SANE doesn't expect users to use xsane.

> Providing a convenient interface for non-programmers probably requires more complex GUI elements and layouts.

Absolutely.

> Me, who does a little programming, I prefer to see the complete config file in whole, opened next to the docs and reload the program only after I have finished editing.

Elektra supports this perfectly.

qwepoizt commented:
Sorry for the delay, here is my answer:

> How does this handle multiple $DISPLAYs?

Thanks for the question! This PR has a limitation regarding the `randr` method of adjusting color temperature: It only affects one `crtc`. This could be fixed rather easily by changing the implementation of [src/gamma-randr.c](https://github.com/jonls/redshift/pull/837/files#diff-2b0175f70e9ba148949f9d3d67d99d8441c800473cf244de9aef72d5ce94b30d).

I am not sure how `$DISPLAY`, `screen`, `crtc`, `output` are related exactly. #23, #26, #183, #242 might be helpful in finding out which of these terms refer to what or to the same concept.

I've updated the top post!
